### ISSUE TITLE
Preparation for V3.1 release

### ISSUE THREAD

**Issue by nudles** [apache/singa#784]
I propose to release a minor version to reflect the changes since v3.0.

Please test the following items and check the documentation if they are done.
  - [x] new onnx models and operators   @joddiy 
  - [x] distributed training @chrishkchris 
  - [x] computational graph to support RNN @chrishkchris 
  - [x] website update @nudles
  - [x] pypi package generation @nudles  
  - [x] github workflow for code quality and coverage management @moazreyad 
  - [x] tensor APIs @dcslin 
  - [x] new autograd operators @joddiy 
  - [ ] anything else?

Here is the checklist and [steps](http://www.apache.org/dev/release-publishing.html)
- [ ] Select a release manager. The release manager (RM) is the coordinator for the release process. It is the RM's signature (.asc) that is uploaded together with the release. The RM generates KEY (RSA 4096-bit) and uploads it to a public key server. The RM needs to get his key endorsed (signed) by other Apache user, to be connected to the web of trust. He should first ask the mentor to help signing his key. [How to generate the key](http://www.apache.org/dev/release-signing.html)?

- [ ] Check license. [FAQ](https://www.apache.org/legal/src-headers.html#faq-docs); [SINGA Issue](https://issues.apache.org/jira/projects/SINGA/issues/SINGA-447)
  - [ ] the codebase does not include third-party code which is not compatible to APL;
  - [ ] The dependencies are compatible with APL. GNU-like licenses are NOT compatible; 
  - [ ] All source files written by us MUST include the Apache license header: http://www.apache.org/legal/src-headers.html. There's a script in there which helps propagating the header to all files.
  - [ ] Update the LICENSE file. If we include any third party code in the release package which is not APL, must state it at the end of the [LICENSE](https://github.com/apache/singa/blob/master/LICENSE#L448) file and include the license boilerplate in the original file.

- [ ] Bump the version. Check code and documentation
  - [ ] The build process is error-free.
  - [ ] Unit tests are included (as much as possible)
  - [ ] Conda packages run without errors. 
  - [ ] The online documentation on the Apache website is up to date.

- [ ] Prepare the RELEASE_NOTES file. Include the following items, Introduction, Features, Bugs (link to JIRA or Github PR), Changes, Dependency list, Incompatibility issues. Follow this [example](http://commons.apache.org/proper/commons-digester/commons-digester-3.0/RELEASE-NOTES.txt). 

- [ ] Prepare DISCLAIMER file. Modify from the [template](http://incubator.apache.org/guides/branding.html#disclaimers)

- [x] Package the release candidate. The release should be packaged into : apache-singa-VERSION.tar.gz. The release should not include any binary files including git files. Upload the release to for [stage](https://dist.apache.org/repos/dist/dev/VERSION/). The tar file, signature, KEY and SHA256 checksum file should be included. MD5 is no longer used. Policy is [here](http://www.apache.org/dev/release-distribution#sigs-and-sums)
    * apache-singa-VERSION.tar.gz
    * KEY
    * XX.acs
    * .SHA256

- [x] Call for vote by sending an email
    ``` 
    To: dev@singa.apache.org
    Subject: [VOTE] Release apache-singa-X.Y.Z (release candidate N)

    Hi all,

    I have created a build for Apache SINGA X.Y.Z, release candidate N.
    The artifacts to be voted on are located here:  xxxx
    The hashes of the artifacts are as follows: xxx
    Release artifacts are signed with the following key: xxx
    Please vote on releasing this package. The vote is open for at least 72 hours and passes if a majority of at least three +1 votes are cast.

   [ ] +1 Release this package as Apache SINGA X.Y.Z
   [ ] 0 I don't feel strongly about it, but I'm okay with the release
   [ ] -1 Do not release this package because...

   Here is my vote:
   +1 
    ```

- [x] Wait at least 48 hours for test responses. Any PMC, committer or contributor can test features for releasing, and feedback. Everyone should check these before vote +1. If the vote passes, then send the result email. Otherwise, repeat from the beginning.
    ```
    Subject: [RESULT] [VOTE] Release apache-singa-X.Y.Z (release candidate N)
    To: dev@singa.apache.org
 
    Thanks to everyone who has voted and given their comments. The tally is as follows.
 
    N binding +1s:
    <names>
 
    N non-binding +1s:
    <names>
 
    No 0s or -1s.
 
     I am delighted to announce that the proposal to release Apache SINGA X.Y.Zhas passed.
     ````

- [ ] Upload the package for [distribution](http://www.apache.org/dev/release-publishing.html#distribution) to https://dist.apache.org/repos/dist/release/VERSION/. 

- [ ] Update the Download page of SINGA website. The tar.gz file MUST be downloaded from mirror, using closer.cgi script; other artifacts MUST be downloaded from main Apache site. More details [here](http://www.apache.org/dev/release-download-pages.html). Some feedback we got during the previous releases:  "Download pages must only link to formal releases, so must not include links to GitHub.",  "Links to KEYS, sigs and hashes must not use dist.apache.org; instead use https://www.apache.org/dist/singa/...;", "Also you only need one KEYS link, and there should be a description of how to use KEYS + sig or hash to verify the downloads."

- [ ] Remove the RC tag and compile the conda packages.

- [ ] Publish the release information. 
    ```
    To: announce@apache.org, dev@singa.apache.org 
    Subject: [ANNOUNCE] Apache SINGA X.Y.Z released

    We are pleased to announce that SINGA X.Y.Z is released.

    SINGA is a general distributed deep learning platform for training big deep learning models over large datasets. 
    The release is available at: http://singa.apache.org/downloads.html
    The main features of this release include XXX
    We look forward to hearing your feedback, suggestions, and contributions to the project.

    On behalf of the SINGA team, {SINGA Team Member Name}
    ````

**Comment by chrishkchris** [comment]
got it, thanks! I will test the distributed training and computational graph to support RNN, also check the corresponding documentation.

**Comment by moazreyad** [comment]
The workflow for coverage is ready in #788 and Travis CI can be removed now as discussed in #790.

More workflows can be made to automate packaging (pypi, debian, etc.) and to automate website updates.

**Comment by nudles** [comment]
Please add details about the changes that you have made since v3.0.0. Follow the [release note](https://github.com/apache/singa/blob/master/RELEASE_NOTES)

**Comment by dcslin** [comment]
changelog 2020-09-24:
- Added
  - BiLSTM model on InsuranceQA example training script.
  - Tensor transformation (reshape, transpose) supports up to 6 dimensions.
  - Implemented traverse_unary_transform in Cuda backend, which is similar to CPP backend one.
  - Added tensor operation erf, rounde (round even).
  - Added some sanity check on autograd input to prevent fatal error caused by unexpected input shape.
- Changed
  - Fix IMDB LSTM model example training script.
  - Fix Tensor operation Mult on Broadcasting use cases.
  - Gaussian function on Tensor now can run on Tensor with odd size.
  - Updated a testing helper function gradients() in autograd to lookup param gradient by param python object id for testing purpose.

**Comment by joddiy** [comment]
- Add new autograd and ONNX operators: Erf, Embedding, Rounde, Round, Where, SpaceToDepth, DepthToSpace, UpSample, Pad, Expand, CosSim, Floor, Ceil.
- Add new layers: Embedding, Gemm.
- Reconstruct `sonnx`:
	- Support creating operators from both layer and autograd.
	- Re-write `SingaRep` to provide more powerful intermediate representation of SINGA.
	- Add a `SONNXModel` which implements from 'Model' to provide uniform API and features.
- Add new ONNX models: SqueezeNet, ShuffleNetv1, ShuffleNetv2, VGG19, DenseNet121, GPT2, RoBERTa.

**Comment by joddiy** [comment]
* 12 new operators are added into the autograd module: Erf, Embedding, Rounde, Round, Where, SpaceToDepth, DepthToSpace, UpSample, Pad, Expand, CosSim, Floor.

  * 2 layers are added into layer module: Embedding, Gemm.

  * 10 new operators are added to sonnx module for both backend and frontend: 
    [Erf](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Erf), 
    [Embedding](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Embedding), 
    [Round](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Round), 
    [Where](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Where), 
    [SpaceToDepth](https://github.com/onnx/onnx/blob/master/docs/Operators.md#SpaceToDepth), 
    [DepthToSpace](https://github.com/onnx/onnx/blob/master/docs/Operators.md#DepthToSpace), 
    [UpSample](https://github.com/onnx/onnx/blob/master/docs/Operators.md#UpSample), 
    [Pad](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Pad), 
    [Expand](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Expand), 
    [Floor](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Floor), 
    Their tests are added as well.

  * Some ONNX models are imported into SINGA, including 
    [SqueezeNet](https://github.com/onnx/models/blob/master/vision/classification/squeezenet), 
    [ShuffleNetv1](https://github.com/onnx/models/blob/master/vision/classification/shufflenet), 
    [ShuffleNetv2](https://github.com/onnx/models/blob/master/vision/classification/shufflenet), 
    [VGG19](https://github.com/onnx/models/blob/master/vision/classification/vgg), 
    [DenseNet121](https://github.com/onnx/models/blob/master/vision/classification/densenet-121), 
    [GPT2](https://github.com/onnx/models/blob/master/text/machine_comprehension/gpt-2), 
    [RoBERTa](https://github.com/onnx/models/blob/master/text/machine_comprehension/roberta),  

  * Reconstruct soonx, 
	- Support creating operators from both layer and autograd.
	- Re-write `SingaRep` to provide more powerful intermediate representation of SINGA.
	- Add a `SONNXModel` which implements from 'Model' to provide uniform API and features.

**Comment by chrishkchris** [comment]
3.1 already released

**Comment by moazreyad** [comment]
> 3.1 already released

I see that there is [vote result email in singa-dev mail list](https://mail-archives.apache.org/mod_mbox/singa-dev/202010.mbox/%3CCAJz0iLunmJ5pqwggCejG2OgzehvLF%3DogEqDFC-kPziV%2BPB%3D37w%40mail.gmail.com%3E). However, I don't see the `Publish the release information.` step. 

An email "[ANNOUNCE] Apache SINGA X.Y.Z released" should be sent to announce@apache.org, dev@singa.apache.org  as given above in the check list of release tasks. This step is required to complete the release process.

Note also that the email body says "The release is available at: http://singa.apache.org/downloads.html", however this page does not exist in the new website. We need to have downloads.html page because it is like a standard in Apache projects to have ProjectName.apache.org/downloads.html . The page `http://singa.apache.org/docs/download-singa/` does not follow the standard URLs of the Apache projects, and it also does not contain the 3.1 release.

Finally, it will be nice to update the [wikipedia page of singa](https://en.wikipedia.org/wiki/Apache_SINGA) with the new version information.

**Comment by chrishkchris** [comment]
Now I added a redirect link so that http://singa.apache.org/downloads.html is working now. It also has the 3.1 release. Does it look better now?

> Note also that the email body says "The release is available at: http://singa.apache.org/downloads.html", however this page does not exist in the new website. We need to have downloads.html page because it is like a standard in Apache projects to have ProjectName.apache.org/downloads.html . The page http://singa.apache.org/docs/download-singa/ does not follow the standard URLs of the Apache projects, and it also does not contain the 3.1 release.

**Comment by nudles** [comment]
I have sent the announcement email.
Thanks for the reminder.

**Comment by moazreyad** [comment]
Great. I updated the infobox of the wikipedia page. I think we can close the issue now.
